docs/design/Path concatenation.md

```diff
@@ -0,0 +1,94 @@
+
+Two _segments_ are equal if all their _chars_ are equal, except for _skip segments_, which are always equal to each other.
+
+DF2 _paths_ are sequences of _segments_ and _separators_. It is illegal for two _segments_ to follow each other. Two _separators_ that have no _chars_ between them are treated as a _separator_-_skip segment_-_separator_ triplet.
+
+Three types of _paths_ represent subsets of the above definition:
+- _Normalized paths_ have no _skip segments_ in them and always have the same number of _separators_ as _segments_. _Normalized paths_ that have at least one _segment_ always start with a _separator_ and end with a _segment_.
+
+[Rationale: it is expected that most implementations will choose to normalize the paths as they enter the process, and work with normalized paths in memory due their superior performance characteristics. The restriction on the start and end of the normalized path are to ensure they always have the same number of _segments_ as they have _separators_, which simplifies the implementation of concatenation functions.]
+
+- _Absolute paths_ are _normalized paths_ that have no _backtrack segments_.
+- _Relative paths_ are _normalized paths_ that can only have _backtrack segments_ at the beginning of the _path_.
+
+[Rationale: the concatenation functions operate on _relative_ and _absolute paths_. Defining them explicitly allows the implementation to not validate the implicit assumption these functions will have to make about their inputs. The restriction of _backtrack segments_ only being legal at the start of a _relative path_ is there to support the simplicity of implementation, both that of the validation and concatenation.]
+
+## `Normalize` specification
+
+`Normalize` is a function with one parameter: `path`, which is a _path_. The result of this function is a _normalized path_. Calling this function for a _path_ is to be called _normalization_, which is to be equivalent to the following algorithm:
+
+1. If the _path_ does not start with a _separator_, prepend one.
+2. If the _path_ ends with a _separator_, delete it.
+3. For each _skip segment_ in the `path`, delete it and the _separator_ preceding it.
+4. Delete all _separator_-_segment_ that is not a _backtrack segment_-_separator_-_backtrack segment_ quadruplets in the `path`.
+5. Repeat step 4 until no such quadruplets are left.
+6. Return the result of the above modifications to `path`.
+
+## `GetFullPath` specification
+
+`GetFullPath` is a function with two parameters: `appendPath`, which is a _relative path_, and `basePath`, which is an _absolute path_. The result of this funtion is an _absolute path_ or an error, whichever is determined by the an algorithm equivalent to the following:
+
+1. Count the number of _backtrack segments_ in `appendPath`.
+2. Delete as many _segments_ in `basePath`, starting from the end. If the operation cannot be performed due to `basePath` not having enough _segments_, return with error.
+3. _Normalize_ `basePath`.
```

Comment:
According to the spec above, absolute paths are already normalized.

Reply:
By this point we have denormalized it:
`basePath` - `/path/to/something/very`
`appendPath` - `/../../important`
1. Count = 1
2. `basePath` = `/path/to//`
3. Here we normalize `basePath` to `/path/to`
4. Here I forgot a step to do the same thing for `appendPath`...
5. Here we concatenate `/path/to` and `/important`.

Will push an update.